larpandoracontent/LArThreeDReco/LArCosmicRay/ThreeViewDeltaRayMatchingAlgorithm.cc

```diff
@@ -0,0 +1,164 @@
+
+    overlapResult = DeltaRayOverlapResult(nMatchedSamplingPoints, nSamplingPoints, chiSquaredSum, xOverlapObject, commonMuonPfoList);
+    
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void ThreeViewDeltaRayMatchingAlgorithm::FindCommonMuonParents(const Cluster *const pClusterU, const Cluster *const pClusterV, const Cluster *const pClusterW,
+    PfoList &commonMuonPfoList) const
+{
+    ClusterList consideredClustersU, consideredClustersV, consideredClustersW;
+    PfoList nearbyMuonPfosU, nearbyMuonPfosV, nearbyMuonPfosW;
+    
+    this->GetNearbyMuonPfos(pClusterU, consideredClustersU, nearbyMuonPfosU);
+
+    if (nearbyMuonPfosU.empty())
+        return;
+    
+    this->GetNearbyMuonPfos(pClusterV, consideredClustersV, nearbyMuonPfosV);
+
+    if (nearbyMuonPfosV.empty())
+        return;
+    
+    this->GetNearbyMuonPfos(pClusterW, consideredClustersW, nearbyMuonPfosW);
+
+    if (nearbyMuonPfosW.empty())
+        return;    
+
+    for (const ParticleFlowObject *const pNearbyMuonU : nearbyMuonPfosU)
+    {
+        for (const ParticleFlowObject *const pNearbyMuonV : nearbyMuonPfosV)
+        {
+            for (const ParticleFlowObject *const pNearbyMuonW : nearbyMuonPfosW)
+            {
+                if ((pNearbyMuonU == pNearbyMuonV) && (pNearbyMuonV == pNearbyMuonW))
+                    commonMuonPfoList.push_back(pNearbyMuonU);
+            }
+        }
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void ThreeViewDeltaRayMatchingAlgorithm::ExamineOverlapContainer()
+{
+    unsigned int repeatCounter(0);
+
+    for (TensorToolVector::const_iterator toolIter = m_algorithmToolVector.begin(); toolIter != m_algorithmToolVector.end(); )
+    {
+        if ((*toolIter)->Run(this, this->GetMatchingControl().GetOverlapTensor()))
+        {
+            toolIter = m_algorithmToolVector.begin();
+
+            if (++repeatCounter > m_nMaxTensorToolRepeats)
+                break;
+        }
+        else
+        {
+            ++toolIter;
+        }
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode ThreeViewDeltaRayMatchingAlgorithm::ReadSettings(const TiXmlHandle xmlHandle)
+{   
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, XmlHelper::ReadValue(xmlHandle, "MuonPfoListName", m_muonPfoListName));
+    
+    AlgorithmToolVector algorithmToolVector;
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, XmlHelper::ProcessAlgorithmToolList(*this, xmlHandle,
+        "DeltaRayTools", algorithmToolVector));
+
+    for (AlgorithmToolVector::const_iterator iter = algorithmToolVector.begin(), iterEnd = algorithmToolVector.end(); iter != iterEnd; ++iter)
```

Comment:
`for (auto tool : algorithmToolVector)`

This is generally preferable if you don't explicitly need the iterator (which usually you don't if you aren't try to manipulate the underlying data structure), as it places the onus to manage comparisons and increments on the compiler rather than the developer